swap/swap.go

```diff
@@ -17,69 +17,340 @@
+		log.Warn(disconnectMessage)
+		return errors.New(disconnectMessage)
+	}
+
+	// calculate new balance
+	var newBalance int64
+	newBalance, err = s.updateBalance(peer.ID(), amount)
+	if err != nil {
+		return
+	}
+
+	// Check if balance with peer crosses the threshold
+	// It is the peer with a negative balance who sends a cheque, thus we check
+	// that the balance is *below* the threshold
+	if newBalance <= -s.paymentThreshold {
+		//if so, send cheque
+		log.Warn(fmt.Sprintf("balance for peer %s went over the payment threshold %v, sending cheque", peer.ID().String(), s.paymentThreshold))
+		err = s.sendCheque(peer.ID())
+		if err != nil {
+			log.Error(fmt.Sprintf("error while sending cheque to peer %s: %s", peer.ID().String(), err.Error()))
+		} else {
+			log.Info(fmt.Sprintf("successfully sent cheque to peer %s", peer.ID().String()))
+		}
+	}
+
+	return
+}
+
+func (s *Swap) updateBalance(peer enode.ID, amount int64) (int64, error) {
 	//adjust the balance
 	//if amount is negative, it will decrease, otherwise increase
-	s.balances[peer.ID()] += amount
+	s.balances[peer] += amount
 	//save the new balance to the state store
-	peerBalance := s.balances[peer.ID()]
-	err = s.stateStore.Put(peer.ID().String(), &peerBalance)
+	peerBalance := s.balances[peer]
+	err := s.stateStore.Put(peer.String(), &peerBalance)
+	if err != nil {
+		log.Error("error while storing balance for peer", "peer", peer.String())
+	}
+	log.Debug("balance for peer after accounting", "peer", peer.String(), "balance", strconv.FormatInt(peerBalance, 10))
+	return peerBalance, err
+}
 
-	log.Debug(fmt.Sprintf("balance for peer %s: %s", peer.ID().String(), strconv.FormatInt(peerBalance, 10)))
+// logBalance is a helper function to log the current balance of a peer
+func (s *Swap) logBalance(peer *protocols.Peer) {
+	err := s.loadState(peer)
+	if err != nil && err != state.ErrNotFound {
+		log.Error(fmt.Sprintf("error while loading balance for peer %s", peer.String()))
+	} else {
+		log.Info(fmt.Sprintf("balance for peer %s is %d", peer.ID(), s.balances[peer.ID()]))
+	}
+}
+
+// sendCheque sends a cheque to peer
+func (s *Swap) sendCheque(peer enode.ID) error {
+	swapPeer := s.getPeer(peer)
+	cheque, err := s.createCheque(peer)
+	if err != nil {
+		log.Error("error while creating cheque: %s", err.Error())
+		return err
+	}
+
+	log.Info(fmt.Sprintf("sending cheque with serial %d, amount %d, benficiary %v, contract %v", cheque.ChequeParams.Serial, cheque.ChequeParams.Amount, cheque.Beneficiary, cheque.Contract))
+	s.cheques[peer] = cheque
+
+	err = s.stateStore.Put(peer.String()+"_cheques", &cheque)
+	// TODO: error handling might be quite more complex
+	if err != nil {
+		log.Error("error while storing the last cheque: %s", err.Error())
+		return err
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// reset balance;
+	// TODO: if sending fails it should actually be roll backed...
+	s.resetBalance(peer, int64(cheque.Amount))
```

Comment:
should not return an error?

Reply:
It should. Not sure about this `TODO`.

Reply:
The TODO means that actually, if the subsequent `swapPeer.Send` would fail, the `resetBalance` should actually be rolled back

Reply:
It now returns an error